### PR TITLE
CMP: fix status held in `OSSL_CMP_CTX` and handling of `failInfo`

### PR DESCRIPTION
**Pull request by DDvO**
So far, there was no clarity at which stage a request went wrong, in particular, already on request construction or during the transaction/transmission.
Moreover, for `genm`/`genp` exchanges the client could not distinguish whether something went wrong or the server responded with an empty list of `InfoTypeAndValue`.

On this occasion, replace magic constants by mnemonic ones.
Update doc accordingly.

Update: also fix handling of unset or missing `failInfo` PKI status information

### CODE REVIEW

**Comment by t8m** [comment]
I am really not sure here, whether this is a bug fix eligible for merging to 3.0 or not.

**Comment by DDvO** [comment]
Asking @openssl/committers for 2nd review.

**Comment by t8m** [comment]
ping for second review

File: crypto/cmp/cmp_client.c
Comment (DDvO) [reply crypto/cmp/cmp_client.c]: Well, these two comments are actually independent of each other.
Yet the (new) first one actually refers to the whole function, so combined it with
```
/* Send certConf for IR, CR or KUR sequences and check response */
```

File: apps/cmp.c
Comment (DDvO) [reply apps/cmp.c]: Indeed this looks confusing.
And the following code block is actually an ideal candidate for getting moved out of of the (anyway overlong) main function into a static helper function. Done so.

File: crypto/cmp/cmp_client.c
Comment (DDvO) [reply crypto/cmp/cmp_client.c]: The new value ` OSSL_CMP_PKISTATUS_request` (-3) is the adequate one.
The former one was ` OSSL_CMP_PKISTATUS_unspecified` (-1).
Yet you're right, better set the value already before calling `ossl_cmp_ctx_set0_newCert()`  - done now.

File: crypto/cmp/cmp_server.c
Comment (DDvO) [reply crypto/cmp/cmp_server.c]: Good point - removed since the constant is now mnemonic :-)

File: crypto/cmp/cmp_server.c
Comment (DDvO) [reply crypto/cmp/cmp_server.c]: Yeah, removed the comment for similar reason, adapting the comment above to
```
    /* determine whether to keep the transaction open or not */
```

File: crypto/cmp/cmp_status.c
Comment (DDvO) [reply crypto/cmp/cmp_status.c]: Done

File: doc/man3/OSSL_CMP_CTX_new.pod
Comment (DDvO) [reply doc/man3/OSSL_CMP_CTX_new.pod]: Yes, this should help - done.

**Comment by DDvO** [comment]
Thanks @tmshort for your review comments.
I've handled them all.

**Comment by DDvO** [comment]
@tmshort, can you approve now?

**Review by tmshort — APPROVED** [review]
Looks good to me. Does @paulidale need to re-approve?

**Comment by DDvO** [comment]
@t8m could you please re-approve after the changes requested by Todd?

**Comment by t8m** [comment]
I think there should be at least some tests added for the issues fixed but it can be a separate PR.

**Comment by DDvO** [comment]
Merged to master and cherry-picked to 3.0 and 3.1.
Thanks @t8m and @tmshort for your reviews.

**Comment by DDvO** [comment]
> I think there should be at least some tests added for the issues fixed but it can be a separate PR.

I've just provided a PR with respective test cases: #19760.

**Pull request closed without merging**
